dhis2/dhis2-core
# feat: add filters to ValidationResult API [DHIS2-6031]

### Summary
Adds 3 filters to `ValidationResultQuery`:

* `pe` to filter on `Period`, allows to give ISO expressions (multiple are combined OR, periods do match on overlap)
* `ou` to filter on `OrganisationUnit` given as UID (multiple are combined OR)
* `vr` to filter on `ValidationRule` given as UID (multiple are combined OR)
* `createdDate` to filter on `Date` the `ValidationResult` was created (all on or after date match)

Filter properties are combined AND.

Filters cannot be used to gain access to otherwise not accessible `ValidationResult`s due to the restrictions added based on the current user.

### Automatic Tests
PR adds 
* unit tests to verify the HQL queries. Most to have quick feedback during development.
* integration tests (database tests) to check the HQL is valid and does actually filter as intended

### Manual Testing
On basis of Sierra Leone sample database:

1. clone fresh database
2. start application
3. goto _Data Quality_ app
4. select _Validation Rule Analysis_
5. change start date to 2020-01-01
6. check _Persist new results_ checkbox
7. select root organisation
8. click _Validate_

To test the API it might be good to add the ALL authority to the superuser role so that no user based filtering occurs that might make the filter results hard to verify.

 1. goto _Users_ app
 2. select _User Roles_
 3. select _Superuser_
 4. in _System_ group check the _All (Full authority)_ checkbox

To test the API use a REST test tool or the browser; some examples:

* http://localhost:8080/api/36/validationResults?ou=jNb63DIHuwU&pe=2020Q1
* http://localhost:8080/api/36/validationResults?ou=byOPfWkK6M6
* http://localhost:8080/api/36/validationResults?vr=ZoG4yXZi3c3
* http://localhost:8080/api/36/validationResults?vr=ZoG4yXZi3c3&createdDate=2021-01-01
